overlord/ifacestate/apparmorprompting/common/patterns.go

```diff
@@ -0,0 +1,478 @@
+	pattern = duplicateSlashes.ReplaceAllString(pattern, `${1}/`)
+	pattern = charsDoublestar.ReplaceAllString(pattern, `${1}*`)
+	pattern = doublestarChars.ReplaceAllString(pattern, `${1}*${2}`)
+	pattern = duplicateDoublestar.ReplaceAllString(pattern, `/**`)
+	pattern = starsAnyMaybeStars.ReplaceAllStringFunc(pattern, func(s string) string {
+		deleteStars := func(r rune) rune {
+			if r == '*' {
+				return -1
+			}
+			return r
+		}
+		return strings.Map(deleteStars, s) + "*"
+	})
+	if strings.HasSuffix(pattern, "/**/*") {
+		// Strip trailing "/*" from suffix
+		return pattern[:len(pattern)-len("/*")]
+	}
+	return pattern
+}
+
+type priorityType int
+
+const (
+	worstPriority priorityType = iota
+	priorityGlobDoublestar
+	priorityTerminalDoublestar
+	priorityDoublestar
+	priorityGlob
+	prioritySinglestar
+	prioritySingleChar
+	priorityTerminated
+	priorityLiteral
+)
+
+type nextPatternsContainer struct {
+	currPriority     priorityType
+	nextPatternsList []*strings.Reader
+}
+
+func (np *nextPatternsContainer) addWithPriority(priority priorityType, reader *strings.Reader) {
+	if priority < np.currPriority {
+		return
+	}
+	if priority > np.currPriority {
+		np.nextPatternsList = np.nextPatternsList[:0]
+		np.currPriority = priority
+	}
+	np.nextPatternsList = append(np.nextPatternsList, reader)
+}
+
+func (np *nextPatternsContainer) nextPatterns() []*strings.Reader {
+	return np.nextPatternsList
+}
+
+// GetHighestPrecedencePattern determines which of the given path patterns is
+// the most specific.
+//
+// Assumes that all of the given patterns satisfy ValidatePathPattern(), so this
+// is not verified as part of this function. Additionally, also assumes that the
+// patterns have been previously expanded using ExpandPathPattern(), so there
+// are no groups in any of the patterns.
+//
+// Below are some sample patterns, in order of precedence, though precedence is
+// only guaranteed between two patterns which may match the same path:
+//
+//	# literals
+//	- /foo/bar/baz
+//	- /foo/bar/
+//	# terminated
+//	- /foo/bar
+//	# any single character
+//	- /foo/bar?baz
+//	- /foo/bar?
+//	- /foo/bar?/
+//	# singlestars
+//	- /foo/bar/*/baz
+//	- /foo/bar/*/
+//	- /foo/bar/*/*baz
+//	- /foo/bar/*/*
+//	- /foo/bar/*
+//	- /foo/bar/*/**
+//	# glob
+//	- /foo/bar*baz
+//	- /foo/bar*/baz
+//	- /foo/bar*/baz/**
+//	- /foo/bar*/
+//	- /foo/bar*/*baz
+//	- /foo/bar*/*/baz
+//	- /foo/bar*/*/
+//	- /foo/bar*/*
+//	- /foo/bar*
+//	# doublestars
+//	- /foo/bar/**/baz
+//	- /foo/bar/**/*baz/
+//	- /foo/bar/**/*baz
+//	# terminal doublestar
+//	- /foo/bar/**/        # These are tough... usually, /foo/bar/**/ would have precedence over
+//	- /foo/bar/**/*       # precedence over /foo/bar/**/*baz, but in this case,
+//	- /foo/bar/**         # the trailing *baz adds more specificity.
+//	# glob with immediate doublestar
+//	- /foo/bar*/**/baz
+//	- /foo/bar*/**/
+//	- /foo/bar*/**
+func GetHighestPrecedencePattern(patterns []string) (string, error) {
```

Comment:
to keep the PR small, I'd move this to a separate one

Reply:
I already split the pattern-related code from the broader `common` package PR #13668, and I think the functions in this PR are related enough that it makes sense for them to be in the same PR. Also, the non-test code is only 478 lines. I suppose it could be justified to make a separate PR for precedence, but it should definitely end up in the same file as the rest of the pattern functions, in my opinion.

Reply:
At the current size splitting is a good idea, this PR as is  definitely still too big. There's no rule that the entire content of a file needs to be added in one PR. Also one doesn't need to understand expansion of {} to review a functions that says it must have been done before.